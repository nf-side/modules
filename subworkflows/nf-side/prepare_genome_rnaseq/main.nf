include { BOWTIE2_BUILD                                         } from '../../../modules/nf-core/bowtie2/build'
include { BOWTIE_BUILD as BOWTIE1_BUILD                         } from '../../../modules/nf-core/bowtie/build'
include { GFFREAD                                               } from '../../../modules/nf-core/gffread'
include { HISAT2_BUILD                                          } from '../../../modules/nf-core/hisat2/build'
include { HISAT2_EXTRACTSPLICESITES                             } from '../../../modules/nf-core/hisat2/extractsplicesites'
include { KALLISTO_INDEX                                        } from '../../../modules/nf-core/kallisto/index'
include { RSEM_PREPAREREFERENCE as MAKE_TRANSCRIPTS_FASTA       } from '../../../modules/nf-core/rsem/preparereference'
include { RSEM_PREPAREREFERENCE as RSEM_PREPAREREFERENCE_GENOME } from '../../../modules/nf-core/rsem/preparereference'
include { SALMON_INDEX                                          } from '../../../modules/nf-core/salmon/index'
include { SAMTOOLS_FAIDX                                        } from '../../../modules/nf-core/samtools/faidx'
include { STAR_GENOMEGENERATE                                   } from '../../../modules/nf-core/star/genomegenerate'

workflow PREPARE_GENOME_RNASEQ {
    take:
    fasta                          // channel: [meta, fasta]
    fasta_fai                      // channel: [meta, fasta_fai]
    gff                            // channel: [meta, gff]
    gtf                            // channel: [meta, gtf]
    splice_sites                   // channel: [meta, splice_sites]
    transcript_fasta               // channel: [meta, transcript_fasta]
    run_bowtie1                    // boolean: true/false
    run_bowtie2                    // boolean: true/false
    run_faidx                      // boolean: true/false
    run_hisat2                     // boolean: true/false
    run_hisat2_extractsplicesites  // boolean: true/false
    run_kallisto                   // boolean: true/false
    run_rsem                       // boolean: true/false
    run_rsem_make_transcript_fasta // boolean: true/false
    run_salmon                     // boolean: true/false
    run_sizes                      // boolean: true/false
    run_star                       // boolean: true/false

    main:
    def join_by_meta_id = { channel1, channel2, channel3 = null ->
        if (channel3) {
            channel1
                .map { meta, content1_ -> [meta.id, content1_, meta] }
                .join(channel2.map { meta, content2_ -> [meta.id, content2_, meta] })
                .join(channel3.map { meta, content3_ -> [meta.id, content3_, meta] })
                .map { _id, content1_, meta1, content2_, meta2, content3_, meta3 -> [meta3 + meta2 + meta1, content1_, content2_, content3_] }
        }
        else {
            channel1
                .map { meta, content1_ -> [meta.id, content1_, meta] }
                .join(channel2.map { meta, content2_ -> [meta.id, content2_, meta] })
                .map { _id, content1_, meta1, content2_, meta2 -> [meta2 + meta1, content1_, content2_] }
        }
    }

    bowtie1_index = Channel.empty()
    bowtie2_index = Channel.empty()
    hisat2_index = Channel.empty()
    kallisto_index = Channel.empty()
    rsem_index = Channel.empty()
    salmon_index = Channel.empty()
    star_index = Channel.empty()
    fasta_sizes = Channel.empty()

    versions = Channel.empty()

    if (run_bowtie1) {
        BOWTIE1_BUILD(fasta)

        bowtie1_index = BOWTIE1_BUILD.out.index
        versions = versions.mix(BOWTIE1_BUILD.out.versions)
    }

    if (run_bowtie2) {
        BOWTIE2_BUILD(fasta)

        bowtie2_index = BOWTIE2_BUILD.out.index
        versions = versions.mix(BOWTIE2_BUILD.out.versions)
    }

    if (run_faidx || run_sizes) {
        SAMTOOLS_FAIDX(fasta.map { meta, fasta_ -> [meta, fasta_, []] }, run_sizes)

        fasta_fai = fasta_fai.mix(SAMTOOLS_FAIDX.out.fai)
        fasta_sizes = SAMTOOLS_FAIDX.out.sizes
        versions = versions.mix(SAMTOOLS_FAIDX.out.versions)
    }

    if (run_hisat2 || run_kallisto || run_rsem || run_rsem_make_transcript_fasta || run_salmon || run_star) {
        GFFREAD(gff.map { meta, gff_ -> [meta, [], gff_] })

        versions = versions.mix(GFFREAD.out.versions)

        gtf = gtf
            .mix(GFFREAD.out.gtf)
            .groupTuple()
            .map { meta, gtf_ -> gtf_[1] ? [meta, gtf_[1]] : [meta, gtf_[0]] }

        if (run_hisat2 || run_hisat2_extractsplicesites) {
            HISAT2_EXTRACTSPLICESITES(gtf)

            versions = versions.mix(HISAT2_EXTRACTSPLICESITES.out.versions)
            splice_sites = splice_sites.mix(HISAT2_EXTRACTSPLICESITES.out.txt)

            if (run_hisat2) {
                HISAT2_BUILD(join_by_meta_id(fasta, gtf, splice_sites))

                hisat2_index = HISAT2_BUILD.out.index

                versions = versions.mix(HISAT2_BUILD.out.versions)
            }
        }

        if (run_kallisto || run_rsem_make_transcript_fasta || run_salmon) {
            MAKE_TRANSCRIPTS_FASTA(join_by_meta_id(fasta, gtf))

            versions = versions.mix(MAKE_TRANSCRIPTS_FASTA.out.versions)

            transcript_fasta = transcript_fasta.mix(MAKE_TRANSCRIPTS_FASTA.out.transcript_fasta)

            if (run_kallisto) {
                KALLISTO_INDEX(transcript_fasta)

                kallisto_index = KALLISTO_INDEX.out.index
                versions = versions.mix(KALLISTO_INDEX.out.versions)
            }

            if (run_salmon) {
                SALMON_INDEX(join_by_meta_id(fasta, transcript_fasta))

                salmon_index = SALMON_INDEX.out.index
                versions = versions.mix(SALMON_INDEX.out.versions)
            }
        }

        if (run_rsem) {
            RSEM_PREPAREREFERENCE_GENOME(join_by_meta_id(fasta, gtf))

            rsem_index = RSEM_PREPAREREFERENCE_GENOME.out.index
            versions = versions.mix(RSEM_PREPAREREFERENCE_GENOME.out.versions)
        }

        if (run_star) {
            STAR_GENOMEGENERATE(join_by_meta_id(fasta, gtf))

            star_index = STAR_GENOMEGENERATE.out.index
            versions = versions.mix(STAR_GENOMEGENERATE.out.versions)
        }
    }

    emit:
    bowtie1_index    // channel: [meta, BowtieIndex/]
    bowtie2_index    // channel: [meta, Bowtie2Index/]
    fasta_fai        // channel: [meta, *.fa(sta).fai]
    fasta_sizes      // channel: [meta, *.fa(sta).sizes]
    gtf              // channel: [meta, gtf]
    hisat2_index     // channel: [meta, Hisat2Index/]
    kallisto_index   // channel: [meta, KallistoIndex]
    rsem_index       // channel: [meta, RSEMIndex/]
    salmon_index     // channel: [meta, SalmonIndex/]
    splice_sites     // channel: [meta, *.splice_sites.txt]
    star_index       // channel: [meta, STARIndex/]
    transcript_fasta // channel: [meta, *.transcripts.fasta]
    versions         // channel: [versions.yml]
}
