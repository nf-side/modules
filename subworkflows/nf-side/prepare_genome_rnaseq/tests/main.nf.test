nextflow_workflow {

    name "Test PREPARE_GENOME_RNASEQ"
    script "../main.nf"
    workflow "PREPARE_GENOME_RNASEQ"

    test(".fasta/.gff - all tools") {

        when {
            params {
            }
            workflow {
                """
                run_bowtie1 = true
                run_bowtie2 = true
                run_faidx = true
                run_hisat2 = true
                run_hisat2_extractsplicesites = true
                run_kallisto = true
                run_rsem = true
                run_rsem_make_transcript_fasta = true
                run_salmon = true
                run_sizes = true
                run_star = true

                input[0] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[1] = Channel.empty()
                input[2] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true)
                ])
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = run_bowtie1
                input[7] = run_bowtie2
                input[8] = run_faidx
                input[9] = run_hisat2
                input[10] = run_hisat2_extractsplicesites
                input[11] = run_kallisto
                input[12] = run_rsem
                input[13] = run_rsem_make_transcript_fasta
                input[14] = run_salmon
                input[15] = run_sizes
                input[16] = run_star
                """
            }
        }

        then {
            def all_files_temp = file(workflow.out.rsem_index.get(0).get(1)).listFiles().sort()
            def rsem_index = all_files_temp.collect { file -> file.name }
            all_files_temp = file(workflow.out.star_index.get(0).get(1)).listFiles().sort()
            def star_index = all_files_temp.collect { file -> file.name }
            all_files_temp = file(workflow.out.salmon_index.get(0).get(1)).listFiles().sort()
            def salmon_index = all_files_temp.collect { file -> file.name }
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.bowtie1_index,
                    workflow.out.bowtie2_index,
                    workflow.out.fasta_fai,
                    workflow.out.fasta_sizes,
                    workflow.out.gtf,
                    workflow.out.hisat2_index,
                    workflow.out.kallisto_index.collect { [it[0], file(it[1]).name] },
                    workflow.out.rsem_index.collect { [it[0], file(it[1]).name + "/", rsem_index] },
                    workflow.out.salmon_index.collect { [it[0], file(it[1]).name + "/", salmon_index] },
                    workflow.out.splice_sites,
                    workflow.out.star_index.collect { [it[0], file(it[1]).name + "/", star_index] },
                    workflow.out.transcript_fasta,
                    workflow.out.versions
                ).match()
                }
            )
        }
    }
}
