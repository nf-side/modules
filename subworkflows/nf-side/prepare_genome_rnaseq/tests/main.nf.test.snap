{
    ".fasta/.gff - all tools": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "test.1.ebwt:md5,e728d922a147241f699fdad1f7ac5a9d",
                        "test.2.ebwt:md5,c0a8d0fffdd7e7a3d9229e9d42a7964a",
                        "test.3.ebwt:md5,123b1717a761500ffa59d54959215d26",
                        "test.4.ebwt:md5,7c9d42e29a60c13a65dd0046a012d848",
                        "test.rev.1.ebwt:md5,dbcbf4caa51963abb964bd59064aeffa",
                        "test.rev.2.ebwt:md5,7cfa0c2adf3488fd184008a7346abba7"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "genome.1.bt2:md5,118f83ac94277f7058d1ecb79fb5acf9",
                        "genome.2.bt2:md5,10a474242f1d77979e1dfc3d862faac8",
                        "genome.3.bt2:md5,123b1717a761500ffa59d54959215d26",
                        "genome.4.bt2:md5,7c9d42e29a60c13a65dd0046a012d848",
                        "genome.rev.1.bt2:md5,21f198b3402e4119a181982ee8b4efd3",
                        "genome.rev.2.bt2:md5,352e3ab4c780f58a7c97f1f6978949a8"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.fasta.fai:md5,3520cd30e1b100e55f578db9c855f685"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.fasta.sizes:md5,b190587cae0531f3cf25552d8aa674db"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.gtf:md5,eb77931882d554f6fcdb2ebbfec062ee"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "genome.1.ht2:md5,8f8a9b1b43511fbc679454cc6f93d203",
                        "genome.2.ht2:md5,afe55bf628b41121cf2c9a999e1dbbcb",
                        "genome.3.ht2:md5,123b1717a761500ffa59d54959215d26",
                        "genome.4.ht2:md5,7c9d42e29a60c13a65dd0046a012d848",
                        "genome.5.ht2:md5,0c5bdeb9777a0bb5c2da117c7b0e2a85",
                        "genome.6.ht2:md5,e055ccdcad39cf608d234ab650eabbb2",
                        "genome.7.ht2:md5,b9e300d0e85cdde63bd6fdcc0d5d5097",
                        "genome.8.ht2:md5,6ae6e44dc59b540e1caadb5b8eeb394a"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "kallisto"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "rsem/",
                    [
                        "Genome",
                        "Log.out",
                        "SA",
                        "SAindex",
                        "chrLength.txt",
                        "chrName.txt",
                        "chrNameLength.txt",
                        "chrStart.txt",
                        "exonGeTrInfo.tab",
                        "exonInfo.tab",
                        "geneInfo.tab",
                        "genome.chrlist",
                        "genome.fasta",
                        "genome.grp",
                        "genome.idx.fa",
                        "genome.n2g.idx.fa",
                        "genome.seq",
                        "genome.ti",
                        "genome.transcripts.fa",
                        "genomeParameters.txt",
                        "sjdbInfo.txt",
                        "sjdbList.fromGTF.out.tab",
                        "sjdbList.out.tab",
                        "transcriptInfo.tab"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "salmon/",
                    [
                        "complete_ref_lens.bin",
                        "ctable.bin",
                        "ctg_offsets.bin",
                        "duplicate_clusters.tsv",
                        "info.json",
                        "mphf.bin",
                        "pos.bin",
                        "pre_indexing.log",
                        "rank.bin",
                        "refAccumLengths.bin",
                        "ref_indexing.log",
                        "reflengths.bin",
                        "refseq.bin",
                        "seq.bin",
                        "versionInfo.json"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test.splice_sites.txt:md5,dc6da761e1dda4b7c490eea01260ee21"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "star/",
                    [
                        "Genome",
                        "Log.out",
                        "SA",
                        "SAindex",
                        "chrLength.txt",
                        "chrName.txt",
                        "chrNameLength.txt",
                        "chrStart.txt",
                        "exonGeTrInfo.tab",
                        "exonInfo.tab",
                        "geneInfo.tab",
                        "genomeParameters.txt",
                        "sjdbInfo.txt",
                        "sjdbList.fromGTF.out.tab",
                        "sjdbList.out.tab",
                        "transcriptInfo.tab"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "genome.transcripts.fa:md5,050c521a2719c2ae48267c1e65218f29"
                ]
            ],
            [
                "versions.yml:md5,2e620478a8847b07ef43e539f3d15a4d",
                "versions.yml:md5,3ef2e2c4f68cff618c1ea7f506974348",
                "versions.yml:md5,8c52d613ef359e337bdfff8893022ff0",
                "versions.yml:md5,8c7e080728ac03ca7d12075b04269c87",
                "versions.yml:md5,9089b357f16db5a1a7f8c19a4a8d347f",
                "versions.yml:md5,9c05faacdd2f90677bfc81c7c70454e7",
                "versions.yml:md5,bdd0980428bddd82b00efe49c7e33203",
                "versions.yml:md5,c082767988a5e7ee97408d28358389aa",
                "versions.yml:md5,c4c0ac84f821dee47cbb5d876778e301",
                "versions.yml:md5,d5b1d7db2c40a4abc8435947fdad1d47",
                "versions.yml:md5,e8962da49ea2a0280fa1ff74caa494c5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.01.0"
        },
        "timestamp": "2025-03-07T15:03:34.583288709"
    }
}